data_safe_haven/administration/users/active_directory_users.py

```diff
@@ -70,11 +66,9 @@ def add(self, new_users: Sequence[ResearchUser]) -> None:
         for line in output.split("\n"):
             self.logger.parse(line)
 
-    def list(self, sre_name: Optional[str] = None) -> Sequence[ResearchUser]:
+    def list(self, sre_name: str | None = None) -> Sequence[ResearchUser]:
```

Comment:
I prefer `Optional[type]` to `type | None` in function signatures because it correctly describes the parameter as being `optional`.

Reply:
[The rule](https://beta.ruff.rs/docs/rules/non-pep604-annotation/) here is enforcing using the new union syntax rather than `typing.Union` or `typing.Optional`. **I don't think it will distinguish when arguments are optional**.

I would go with the rule (apart from where it breaks Typer) because

- It is more concise
- It is more clear that the type is actually a union of T and None
- An `Optional[T]` is not the same as an optional argument, _e.g._

```Python
def f(x: Optional[int]) -> None:
    if x is None:
        do_something()
    else:
        do_something_else()
```
x is `Optional[int]` but it is not an optional argument.

Reply:
I mean that I like the use of `Optional[type]` to indicate that you *do have* an optional argument. It's not a hill I would die on though!

Reply:
Yep. The linter doesn't distinguish between using `typing.Optional` for optional arguments and in other places. It just prefers the newer union operator syntax. So I don't think there is much hope of enforcing that.

I'll keep the rule here apart from exclusions for Typer.